posthog/models/action.py

```diff
@@ -36,49 +37,42 @@ def calculate_events(self, start=None, end=None):
                 event_query, params = (
                     Event.objects.query_db_by_action(self, start=start, end=end).only("pk").query.sql_with_params()
                 )
-
         except EmptyResultSet:
-            self.last_calculated_at = calculated_at
-            self.is_calculating = False
-            self.save()
             self.events.all().delete()
-            return
-
-        query = """DELETE FROM "posthog_action_events" WHERE "action_id" = {}""".format(self.pk)
-
-        period_delete_query = """AND "event_id" in
-                       (SELECT id
-                        FROM posthog_event
-                        WHERE "created_at" >= '{}'
-                        AND "created_at" < '{}');
-                """.format(
-            start.isoformat(), end.isoformat()
-        )
-
-        insert_query = """INSERT INTO "posthog_action_events" ("action_id", "event_id")
-                        {}
-                    ON CONFLICT DO NOTHING
-                 """.format(
-            event_query.replace("SELECT ", "SELECT {}, ".format(self.pk), 1)
-        )
-
-        if not recalculate_all:
-            query += period_delete_query
         else:
-            query += ";"
-
-        query += insert_query
-
-        cursor = connection.cursor()
-        with transaction.atomic():
-            try:
-                cursor.execute(query, params)
-            except Exception as err:
-                capture_exception(err)
-
-        self.is_calculating = False
-        self.last_calculated_at = calculated_at
-        self.save()
+            delete_query = f"""DELETE FROM "posthog_action_events" WHERE "action_id" = {self.pk}"""
+
+            if not recalculate_all:
+                delete_query += f""" AND "event_id" IN (
```

Comment:
My worry here is that this will fall over hard on any larger postgres instance.

What does the query plan look like for this subquery?

Reply:
Thinking about it more - do we really need this delete query if we're not recalculating all? 

The below insert query should handle properly the case where we calculate the same event multiple times and if action changes that should trigger a full recalculation.

Reply:
It should not fall over, in fact this has been our recommended approach for larges Postgres instances, since on-the-fly was hammering Redis and Postgres really hard. But TBH it's difficult to analyze a query plan here, as this is very different depending on what actions an instance has defined.

I kind of agree that this should be practically the same without the `DELETE`, but I was removed from the initial implementation of this periodic mapping approach so there may be some context I'm missing. 🤔

Reply:
I'm not sure I understood the above. _what_ is the recommended approach for Postgres instances?

--- 

The original comments where specifically about doing the DELETE when not doing a full recalculate. This seems bad due to:
1. *Performance* (1) - Doing an event table query (without a team_id filter) is likely to be slow on larger instances. This could easily take minutes if the query planner decides to do a full table scan (hence the question on query plans)
2. *Performance* (2) - Deletes result in trashing in the table and extra vacuuming needs to be done on the postgres instance to clear things up.
3. *Correctness* - Given that the below INSERT statement ignores duplicate rows (via the ON CONFLICT clause), I don't think this is needed for the query to be correct. 

I'm not 100% sure on point (3) this though - is there a hidden reason why we need to delete data every 30s for this to work?

Reply:
Looking at this one more time I see I missed things on my first read - this query was as such previously as well. I still wonder about (3) though - why is it needed?